scripts/build/check_init_priorities.py

```diff
@@ -0,0 +1,316 @@
+import pathlib
+import pickle
+import sys
+
+from elftools.elf.elffile import ELFFile
+from elftools.elf.relocation import RelocationSection
+from elftools.elf.sections import SymbolTableSection
+
+# This is needed to load edt.pickle files.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..",
+                                "dts", "python-devicetree", "src"))
+from devicetree import edtlib  # pylint: disable=unused-import
+
+# Prefix used for relocation sections containing initialization data, as in
+# sequence of "struct init_entry".
+_INIT_SECTION_PREFIX = (".rel.z_init_", ".rela.z_init_")
+
+# Prefix used for "struct device" reference initialized based on devicetree
+# entries with a known ordinal.
+_DEVICE_ORD_PREFIX = "__device_dts_ord_"
+
+# File name suffix for object files to be scanned.
+_OBJ_FILE_SUFFIX = ".c.obj"
+
+# Defined init level in order of priority.
+_DEVICE_INIT_LEVELS = ["EARLY", "PRE_KERNEL_1", "PRE_KERNEL_2", "POST_KERNEL",
+                      "APPLICATION", "SMP"]
+
+# File name to check for detecting and skiping nested build directories.
+_BUILD_DIR_DETECT_FILE = "CMakeCache.txt"
+
+# List of compatibles for node where the initialization priority should be the
+# opposite of the device tree inferred dependency.
+_INVERTED_PRIORITY_COMPATIBLES = set()
+
+class Priority:
+    """Parses and holds a device initialization priority.
+
+    Parses an ELF section name for the corresponding initialization level and
+    priority, for example ".rel.z_init_PRE_KERNEL_155_" for "PRE_KERNEL_1 55".
+
+    The object can be used for comparing levels with one another.
+
+    Attributes:
+        name: the section name
+    """
+    def __init__(self, name):
+        for idx, level in enumerate(_DEVICE_INIT_LEVELS):
+            if level in name:
+                _, priority = name.strip("_").split(level)
+                self._level = idx
+                self._priority = int(priority)
+                self._level_priority = self._level * 100 + self._priority
+                return
+
+        raise ValueError("Unknown level in %s" % name)
+
+    def __repr__(self):
+        return "<Priority %s %d>" % (
+                _DEVICE_INIT_LEVELS[self._level], self._priority)
+
+    def __str__(self):
+        return "%s %d" % (_DEVICE_INIT_LEVELS[self._level], self._priority)
+
+    def __lt__(self, other):
+        return self._level_priority < other._level_priority
+
+    def __eq__(self, other):
+        return self._level_priority == other._level_priority
+
+    def __hash__(self):
+        return self._level_priority
+
+
+class ZephyrObjectFile:
+    """Load an object file and finds the device defined within it.
+
+    Load an object file and scans the relocation sections looking for the known
+    ones containing initialization callbacks. Then finds what device ordinals
+    are being initialized at which priority and stores the list internally.
+
+    Attributes:
+        file_path: path of the file to be loaded.
```

Comment:
`defined_devices` isn't documented

Reply:
Added a paragraph about it, lmk if it's not the correct format.